### PR TITLE
perf(core): reduce persistent meta when delete large scale segment

### PR DESCRIPTION
**Pull request by lifepuzzlefun**

### CODE REVIEW

**Comment by lifepuzzlefun** [comment]
log 
```java
[2024-07-22 10:30:21,309] ERROR Uncaught exception in scheduled task 'delete-file' (org.apache.kafka.server.util.KafkaScheduler)
java.lang.OutOfMemoryError: Java heap space
        at com.fasterxml.jackson.core.util.ByteArrayBuilder.toByteArray(ByteArrayBuilder.java:163)
        at com.fasterxml.jackson.databind.ObjectWriter.writeValueAsBytes(ObjectWriter.java:1164)
        at kafka.log.streamaspect.ElasticLogMeta.encode(ElasticLogMeta.java:53)
        at kafka.log.streamaspect.ElasticLogSegmentManager.asyncPersistLogMeta(ElasticLogSegmentManager.java:80)
        at kafka.log.streamaspect.ElasticLogSegmentManager$EventListener.onEvent(ElasticLogSegmentManager.java:153)
        at kafka.log.streamaspect.ElasticLogSegment.deleteIfExists(ElasticLogSegment.java:410)
        at kafka.log.LocalLog$.$anonfun$deleteSegmentFiles$5(LocalLog.scala:950)
        at kafka.log.LocalLog$.$anonfun$deleteSegmentFiles$5$adapted(LocalLog.scala:949)
        at kafka.log.LocalLog$$$Lambda$5148/0x0000000801ba09c0.apply(Unknown Source)
        at scala.collection.immutable.List.foreach(List.scala:334)
        at kafka.log.LocalLog$.$anonfun$deleteSegmentFiles$4(LocalLog.scala:949)
        at kafka.log.LocalLog$.deleteSegments$1(LocalLog.scala:739)
        at kafka.log.LocalLog$.$anonfun$deleteSegmentFiles$6(LocalLog.scala:956)
        at kafka.log.LocalLog$$$Lambda$5145/0x0000000801455d88.run(Unknown Source)
        at org.apache.kafka.server.util.KafkaScheduler.lambda$schedule$1(KafkaScheduler.java:150)
        at org.apache.kafka.server.util.KafkaScheduler$$Lambda$1237/0x000000080155c7c8.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by lifepuzzlefun** [comment]
this happens when 1gb/s retention time from 3 days change to 12h. the partition number is 100.
one of broker OOM the delete segment in one time is very very big. and the OOM cause the kraft ping packet will be error and the broker remain in fence can't recover

after the delete change from 200TB to 18TB

File: core/src/main/scala/kafka/log/LocalLog.scala [hunk 1]
```diff
@@ -946,9 +947,19 @@ object LocalLog extends Logging {
       info(s"${logPrefix}Deleting segment files ${segmentsToDelete.mkString(",")}")
```
Comment (lifepuzzlefun) [inline core/src/main/scala/kafka/log/LocalLog.scala:947]: this log will also print a lot elastic log inner state. but not changed in this pr

**Comment by lifepuzzlefun** [comment]
@superhx would you check this ?

**Comment by superhx** [comment]
> this cause one of broker OOM when 1gb/s topic retention time change from 3 days to 12h. after the delete change from 200TB to 18TB. the partition number is 100. the delete segment at the same time is very very big.
> 
> and the OOM cause the kraft ping packet will be error and the broker remain in fence can't recover

Only trigger the last deleted segment deletion is little bit tricky.

Perhaps we can solve this problem by internally batch the ElasticLogSegmentManager$EventListener#onEvent. Let its effect become that even if 100 segments are deleted in sequence, it will only trigger two saves in the end.

**Comment by lifepuzzlefun** [comment]
new PR #1670

**Pull request closed without merging**